### PR TITLE
feat: store results to database and UI updates

### LINKED ISSUE #892: capa explorer install note: $IDAUSR/plugins
add note of `$IDAUSR/plugins` as per https://hex-rays.com/blog/igors-tip-of-the-week-33-idas-user-directory-idausr/

### LINKED ISSUE #1279: IDA explorer: persist results across runs
Is it worth storing results once the plugin ran so we can load it without doing the same work again in a subsequent run?

### PR DESCRIPTION
**Pull request by mr-tz**
still a WIP, but looking for feedback/testing

likely pending some updates from #1291 wrt rule caching/saving

closes #1279 
closes #892

2023-01-20_15-12-58_ida
2023-01-20_15-13-05_ida

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: CHANGELOG.md
Comment (williballenthin) [inline CHANGELOG.md:101]:
```suggestion
- cache capa results across IDA sessions #1279 @mr-tz
```

File: capa/ida/helpers.py
Comment (williballenthin) [inline capa/ida/helpers.py:230]: recommend naming these routines as `load/save_cached_results`?

File: capa/ida/helpers.py
Comment (williballenthin) [inline capa/ida/helpers.py:223]: we'll have to be careful with the interplay between the capa ruleset cache and the explorer ruleset cache

File: capa/ida/plugin/form.py
Comment (williballenthin) [inline capa/ida/plugin/form.py:157]: i like that this is disabled when there's nothing to do.

**Review by williballenthin — COMMENTED** [review]
I like the functionality that this PR introduces.

I think I'm most concerned about loading staling results or rulesets. Right now, it doesn't seem like there's any checking or validation. I'm afraid this might lead to confusing behavior, such as a user is developing a rule, changing its content, and doesn't see the updated logic. I guess this only applies if the user were to be re-opening IDA often, but still...

Could we introduce any metadata to make it easier to detect weird situations? For example:
  - to detect stale rules, hash the rule content and/or path
  - to detect stale analysis, list the list of all addresses in the IDB (to try to catch rebase/new segments)

I'm open to discussion if you think this isn't needed, but in either case, lets document the behavior in comments.

File: capa/ida/helpers.py
Comment (williballenthin) [inline capa/ida/helpers.py:206]: we may want to introduce a version here, in the payload blobs, or even better, in the netnode path, so that we if we introduce a new format for the cached data, we don't run into trouble.

File: capa/ida/helpers.py
Comment (williballenthin) [inline capa/ida/helpers.py:40]:
```suggestion
CAPA_NETNODE = "$ com.mandiant.capa.v1"
```

see below discussion on versioning

Comment (williballenthin) [reply capa/ida/helpers.py:40]: and/or use `capa.version.__version__` directly in the path, and then we can rely on each release of capa to be namespaced.

File: capa/ida/helpers.py
Comment (williballenthin) [reply capa/ida/helpers.py:223]: maybe we should just rely on the disk cached ruleset? its pretty fast and doing this would reduce the chance of weird staleness issues.  just an idea, open to discussion.

Comment (mr-tz) [reply capa/ida/helpers.py:223]: yes, wanted to wait finalizing this after #1291 is complete

**Comment by mr-tz** [comment]
Thanks for the comments and suggestions. Once #1291 is done I'll rebase and address your feedback.

File: capa/ida/helpers.py
Comment (mr-tz) [reply capa/ida/helpers.py:223]: > maybe we should just rely on the disk cached ruleset
yes, I think the changes in the other PR lead to that, any way

**Comment by mr-tz** [comment]
New UI:
2023-01-26_12-24-07_ida

Settings:
2023-01-26_12-24-40_ida

On rebase / load with invalid address vs. what's in the results:
```
Cached results appear to be invalid, please reanalyze program. Please refer to IDA Output window for more information.
ERROR:capa:0x402860 is not a valid location in this database
```

File: capa/ida/helpers.py
Comment (mr-tz) [reply capa/ida/helpers.py]: removed here, now only using system rule cache

File: capa/ida/helpers.py
Comment (mr-tz) [inline capa/ida/helpers.py:229]: This should ensure all result addresses exist in the database.

**Review by mike-hunhoff — CHANGES_REQUESTED** [review]
This is awesome @mr-tz! I have a few comments below on broken workflows that we should address before merging. Also, in general, there is noticeable delay when clicking buttons/boxes including `Analyze/Load`, `Settings`, `Limit results to current function`, and `Show matches by function`. Is there any way to improve this? For reference, I tested using the mimikatz example binary.

### Broken workflows
- [ ] Noticeable delay when checking `Show matches by function` before a "Please confirm" modal opens asking to load existing data or analyze program again
  - assuming delay is result of verifying cached results, ideally this only happens when the plugin is first opened; selecting `Limit results to current function` and `Show matches by function` should not incur additional overhead
- [ ] selecting "Analyze on plugin start (ask before...)" results in "capa rules..." label in bottom left of UI not being populated at startup
- [ ] with stale results, selecting `Show matches by function` opens modal asking "This database contains capa results generated..."; selecting `Loading existing results` results in empty Program Analysis window; selecting `Reanalyze program` does not kick off analysis as expected
  - does it make sense to store the program base for the cached results and, if it changes, force the user to rerun analysis at startup? Rebasing invalidates cached results to a point where **not** forcing the user to rerun analysis will likely cause confusion
- [ ] when using rule generator "loaded existing capa results from database" does not change even though the rule generator always pulls fresh results; this could be confusing to users
  - can we remove/add/update this message when switching views?

### General suggestions
- [ ] noticeable delay when rendering cached results with no notification to the user
  - when analyzing fresh we use the progress modal to inform users of capa's hard work; can we do similar when loading the cached results?
- [ ] do not make it the default to start analysis as soon as the plugin is opened; this took me by surprise being used to kicking off the analysis manually, and isn't great for samples that take along time to anlayze
- [ ] leave the `Analyze/Load` button as `Analyze` as we prompt the user with additional information about loading cached results; we use the same button for the rule generator which could cause confusion as we always run a fresh analysis

File: capa/ida/helpers.py
Comment (williballenthin) [inline capa/ida/helpers.py:40]: 👍🏼

File: capa/ida/plugin/README.md
Comment (williballenthin) [inline capa/ida/plugin/README.md:52]:
```suggestion
   - common paths are `%APPDATA%\Hex-Rays\IDA Pro\plugins` (Windows) or `$HOME/.idapro/plugins` on Linux/Mac
```

**Review by williballenthin — APPROVED** [review]
lgtm, but i've been much less thorough than @mike-hunhoff so please address his review first.

File: capa/ida/plugin/README.md
Comment (mr-tz) [reply capa/ida/plugin/README.md:52]: nice catch

File: capa/ida/helpers.py
Comment (mike-hunhoff) [inline capa/ida/helpers.py:207]: Does this create `CAPA_NETNODE` if it does not already exist? Is `n` always valid in this context?

File: capa/ida/helpers.py
Comment (mike-hunhoff) [inline capa/ida/helpers.py:219]: Change to `verify_and_load_cached_results`?

File: capa/ida/plugin/form.py
Comment (mike-hunhoff) [inline capa/ida/plugin/form.py:291]: What about when user navigates from program analysis to rule generator and back?

File: capa/ida/plugin/form.py
Comment (mike-hunhoff) [inline capa/ida/plugin/form.py:333]: We overload the reset button for both tabs and it completes different tasks depending on the selected tab. This is another instance calling for us to rework the buttons so it's less confusing for the user.

File: capa/ida/plugin/form.py
Comment (mike-hunhoff) [inline capa/ida/plugin/form.py:665]: - My first thought is **not** to keep the cached rule set and cached results in sync
    - Suggested workflow:
      - Check for cached results
        - If found, load cached results ignoring cached rule set
          - Update bottom left status message to say "Loaded existing capa results from..."
          - Note: this should also save use CPU cycles by not loading the cached rule set when we don't use it
        - Else, load cached rule set, run anlaysis, cache results
          - Update bottom left status message to say "capa rules: <rule_path> (<rule_count> rules)"

Use-case: user creates a new rule using the rule generator and adds it to their rule path. User is now forced to re-analyze even if they do not want to spend time waiting for the new rule to be included.

Open to discussion here as I see benefits to both approaches.

**Review by mike-hunhoff — CHANGES_REQUESTED** [review]
Huge improvement, thank you @mr-tz ! I left some comments to address as well as the following:

Note: we have been overloading the buttons and status message to serve both the program analysis and rule generator tabs. Adding this caching feature is making me realize that the overloading has probably reached a point where it is more confusing than simply adding separate buttons and status message to each tab view. I think we should focus on getting the caching right in this PR and not so much about keeping the buttons and status messages in sync. Once we merge these changes we can open a follow-up PR to address the buttons and status messages.

- [ ] `anlayze` -> `load existing results` - does not populate "capa rules: ..." status message in bottom left; proceeding to select "Show matches by function" does populate "capa rules: ..." in bottom left
- [ ] `anlayze` -> `cancel` clears status message in lower left; ideally the status message stays the same and does not dissaper if the user selects `cancel`
- [ ] switching tabs clears status message in lower left; ideally this does not disappear

File: capa/ida/plugin/form.py
Comment (mr-tz) [reply capa/ida/plugin/form.py:291]: moving this to #1305, I think it's fine as is, but can be improved

File: capa/ida/plugin/form.py
Comment (mr-tz) [reply capa/ida/plugin/form.py:665]: Changed this to warn the user, but ultimately letting them decide to leave or reanalyze. What do you think of this?

File: capa/ida/helpers.py
Comment (mr-tz) [reply capa/ida/helpers.py:207]: Yes, it creates a new one or accesses the existing one.

File: capa/ida/plugin/form.py
Comment (mr-tz) [reply capa/ida/plugin/form.py]: 2023-02-03_11-06-03_ida

**Review by mike-hunhoff — APPROVED** [review]
@mr-tz your most recent changes are great! I pushed a few commits to this PR to improve the buttons, status, and error messages. I think the overall flow of the caching mechanism is super smooth. Pending your final review/thoughts LGTM!

TLDR; my changes ensure the UI buttons make sense for the focused tab and that the status message is cached when switching between views.

Screen Shot 2023-02-03 at 2 45 39 PM
Screen Shot 2023-02-03 at 2 45 21 PM
Screen Shot 2023-02-03 at 2 44 34 PM

File: capa/ida/plugin/form.py
Comment (mr-tz) [inline capa/ida/plugin/form.py:893]:
```suggestion
```

**Review by mr-tz — COMMENTED** [review]
nice simplifications and improvements!

File: capa/ida/plugin/form.py
Comment (mike-hunhoff) [reply capa/ida/plugin/form.py:893]: good catch!

**Comment by mike-hunhoff** [comment]
> nice simplifications and improvements! seems like this closes #1305 now?!

yes it does! LGTM!!

**Pull request merged**
